go/apps/mixnet/queue.go

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2015, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mixnet
+
+import (
+	"container/list"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+)
+
+type sendQueueError struct {
+	id uint64 // Serial identifier of sender to which error pertains.
+	error
+}
+
+// The SendQueue structure.
+type SendQueue struct {
+	batchSize, ct int
+
+	network string
+
+	nextAddr   map[uint64]string
+	nextConn   map[uint64]net.Conn
+	sendBuffer map[uint64]*list.List
+
+	enqueue chan Queueable
+	err     chan sendQueueError
+}
+
+// NewSendQueue creates a new SendQueue structure.
+func NewSendQueue(network string, batchSize int) (sq *SendQueue) {
+	sq = new(SendQueue)
+	sq.batchSize = batchSize
+	sq.network = network
+
+	sq.nextAddr = make(map[uint64]string)
+	sq.nextConn = make(map[uint64]net.Conn)
+	sq.sendBuffer = make(map[uint64]*list.List)
+
+	sq.enqueue = make(chan Queueable)
+	sq.err = make(chan sendQueueError)
+	return sq
+}
+
+// Enqueue adds a queueable object, such as a message or directive, to the
+// send queue.
+func (sq *SendQueue) Enqueue(q *Queueable) {
+	sq.enqueue <- *q
+}
+
+// DoSendQueue atomically adds messages to a queue and transmits messages in
```

Comment:
I don't see any synchronization in this function. Do you expect in the future to run this concurrently? If not, maybe remove "atomically" here, since it won't be meaningful.